### PR TITLE
Make installer register triplea as handler for triplea: domains (Partly #1723)

### PR DESCRIPTION
**Pull request by RoiEXLab**
This adds the capability to handle custom urls by triplea for Windows (tested on Windows 10) and common Linux Systems (all freedesktop.org based systems which are using) (tested on XUbuntu 16.04.2 LTS).
OS X support would be great, but I don't own a mac so I wouldn't be able to test stuff. Also I have no idea on how to achieve this on mac systems 🤔 The Internet has very little information about this, compared to other oses.

Side Effect/Good News: I enabled LZMA compression and increased the compression level of the installer's executable. The files are now 15MB smaller 🎉 

Disclaimer: This PR only updates the installer, the necessary files are created when installing and completely removed when uninstalling. Another PR will make triplea handle `triplea:*` domains.

#### Details
Typing `triplea:<somestring>` into your browser causes it to call triplea like this: `/path/to/triplea triplea:<somestring>`.
2 Questions:

- Should we go for a flag like `-o`? in this case we'd call triplea like this: `/path/to/triplea -o triplea:<somestring>`
- What are we going to pass in as params? just the maps name? The full URL? Both?

@ron-murhammer @DanVanAtta @ssoloff @simon33-2

### CODE REVIEW

File: build.install4j [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<install4j version="6.1.2" transformSequenceNumber="5">
+<install4j version="6.1.5" transformSequenceNumber="5">
```
Comment (ssoloff) [inline build.install4j]: @RoiEXLab I'll do a detailed review later, but FYI...  The Travis build currently uses install4j 6.1.2:

```
[WARNING] The version 6.1.2 of this install4j installation is lower than the version 6.1.5 of the IDE that saved the config file.
```

It's funny that you submitted this now because I am preparing some PRs to upgrade install4j to 6.1.5.  However, it will require multiple PRs across multiple repos (the install4j installer is located in the `triplea-game/assets` repo), so it may not happen quickly.

You may want to side-by-side install install4j 6.1.2 on your system and point your `install4jHomeDir` project variable there to verify your changes work as expected when building using that version.  If this PR gets merged before the install4j 6.1.5 upgrade PRs, I'll make sure to update this file as needed.

Comment (RoiEXLab) [inline build.install4j]: I tested building the files using your updated build.gradle file so we should be fine...
I built the files using gradlew anyways, since there is no other way unless changing the install4j file

Comment (ssoloff) [inline build.install4j]: That change only updates the install4j Gradle plugin--it does not update install4j itself.

The Gradle build uses whatever install4j installation your `install4jHomeDir` project variable is set to.  In the Travis build, it points to install4j 6.1.2 downloaded and installed from the `triplea-game/assets` repo.  See [`.travis/setup`](https://github.com/triplea-game/triplea/blob/master/.travis/setup) for the details.

I'm going to submit the PR to put install4j 6.1.5 into the `assets` repo now.  We may just be able to use this PR to upgrade the entire build to 6.1.5.  That way we don't have to do 6.1.5 testing twice. :smiley:

**Comment by ron-murhammer** [comment]
@ssoloff @simon33-2 I'd appreciate if one of you can do a review of this one. Then I'll just do a quick review once you approve of it.

**Comment by ssoloff** [comment]
I'd be happy to shepherd this one through as I'd like to see if we can do the install4j upgrade at the same time (since @RoiEXLab already did his development with 6.1.5).  I believe @simon33-2 has tested the OSX installer before, so that would be invaluable if he can do it again to ensure no regressions are introduced for Mac.

@ron-murhammer triplea-game/assets#10 is a prerequisite if we are going to upgrade install4j in this PR.  Do you mind merging it when you have a chance?  It simply adds the install4j 6.1.5 installer to the assets repo.

**Comment by ron-murhammer** [comment]
@ssoloff Done. Sounds good to me.

**Comment by RoiEXLab** [comment]
@ssoloff To give you a basic understanding of what I did:
For Windows I added a couple actions which write specific keys into the registry.
For Linux I added some java code which replaces specific strings in some configuration files.
To ensure only Linux systems are affected by this everything is nested inside a big if Útil.isLinux.
Of course I added scripts for the uninstaller as well in order to remove any values afterwards.
So reviewing the java code snippet would be the biggest task.

**Comment by ssoloff** [comment]
@RoiEXLab Thanks for the summary.  I plan on doing a quick review before I go offline for the night, but should be able to fully review it within the next 24 hours.  I also have some test cases I want to run through.

In the meantime, I've attached a [patch](https://github.com/triplea-game/triplea/files/1043507/upgrade-install4j-6.1.5.diff.txt) that will upgrade the build to install4j 6.1.5.  It should apply cleanly to this branch after you merge the latest changes in from `master`.  I already verified that it works on [Travis](https://travis-ci.org/ssoloff/triplea/builds/238185888).

File: build.install4j [hunk 2]
```diff
@@ -334,6 +334,177 @@
...
+                              </void>
+                            </object>
+                          </void>
+                        </object>
+                      </java>
+                    </serializedBean>
+                    <condition />
+                  </action>
+                </beans>
+              </group>
```
Comment (ssoloff) [inline build.install4j]: A general question about the new Linux install/uninstall actions...  Will they handle the case where I install multiple versions of TripleA side-by-side?  I don't know about others, but I do this regularly.

My expectation would be that the `triplea` protocol handler would always point to the latest version.  For example:

* If I have version 1 already installed and then install version 2 to a different folder, I would expect the protocol handler to be updated to version 2.
* If I have version 2 already installed and then install version 1 to a different folder (say to try to reproduce a bug), I would expect the protocol handler to remain unchanged.

Note that this could severely complicate the uninstall logic if it has to restore the protocol handlers to point to the previous version instead of just removing them.

What are your thoughts on this, @RoiEXLab?  Are side-by-side installations something we need to worry about for a general user?

File: build.install4j [hunk 3]
```diff
@@ -334,6 +334,177 @@
...
+                String nextLine = scanner.nextLine();
+                if(nextLine.startsWith("Exec=") &amp;&amp; !nextLine.contains("%u")){
+                    builder.append(nextLine).append(" \"%u\"\n");
+                } else if(!nextLine.startsWith(mimeTypePrefix)){
+                    builder.append(nextLine).append('\n');
+                }
+            }
+            builder.append(mimeTypePrefix).append(mimeType);
+        }
+        try(PrintWriter writer = new PrintWriter(desktopFile)){
```
Comment (RoiEXLab) [inline build.install4j]: No, because install4j removes the desktop file automatically.

File: build.install4j [hunk 4]
```diff
@@ -45,7 +45,7 @@
     <components />
   </files>
   <launchers>
-    <launcher name="TripleA" id="33" customizedId="" external="false" excludeFromMenu="false" unixMode="755" menuName="TripleA-${compiler:sys.version}" icnsFile="" customMacBundleIdentifier="false" macBundleIdentifier="" swtApp="false" fileset="" macBundleBinary="JavaApplicationStub" addMacEntitlements="false" macEntitlementsFile="" useCustomMacosExecutableName="false" customMacosExecutableName="" useJavaMinVersionOverride="false" javaMinVersionOverride="" useJavaMaxVersionOverride="false" javaMaxVersionOverride="">
+    <launcher name="TripleA" id="33" customizedId="" external="false" excludeFromMenu="false" unixMode="755" unixAutoStart="true" menuName="TripleA-${compiler:sys.version}" icnsFile="" customMacBundleIdentifier="false" macBundleIdentifier="" swtApp="false" fileset="" macBundleBinary="JavaApplicationStub" addMacEntitlements="false" macEntitlementsFile="" useCustomMacosExecutableName="false" customMacosExecutableName="" useJavaMinVersionOverride="false" javaMinVersionOverride="" useJavaMaxVersionOverride="false" javaMaxVersionOverride="">
```
Comment (RoiEXLab) [inline build.install4j]: Good question... I'm editing the file using the gui editor I'm not sure what caused this

File: build.install4j (hunk 2 above)
Comment (RoiEXLab) [inline build.install4j]: Good point.
While currently on Windows old handlers are overridden properly (I just noticed this may not always be working on linux)
Handlers are completely when uninstalling any version.
I will fix that. Shouldn't be too difficult.

File: build.install4j [hunk 5]
```diff
@@ -417,7 +588,7 @@
           </screen>
         </screens>
       </application>
-      <application name="" id="uninstaller" customizedId="" beanClass="com.install4j.runtime.beans.applications.UninstallerApplication" enabled="true" commentSet="false" comment="" actionElevationType="none" fileset="" customIcnsFile="" customIcoFile="" macEntitlementsFile="" automaticLauncherIntegration="false" launchMode="startupFirstWindow" launchInNewProcess="false" launchSchedule="updateSchedule" allLaunchers="true">
+      <application name="" id="uninstaller" customizedId="" beanClass="com.install4j.runtime.beans.applications.UninstallerApplication" enabled="true" commentSet="false" comment="" actionElevationType="elevated" fileset="" customIcnsFile="" customIcoFile="" macEntitlementsFile="" automaticLauncherIntegration="false" launchMode="startupFirstWindow" launchInNewProcess="false" launchSchedule="updateSchedule" allLaunchers="true">
```
Comment (RoiEXLab) [inline build.install4j]: I'm not sure if this is even necessary.
Uninstalling wasn't working so I tried stuff like this.
I will check if it works without it.

File: build.install4j (hunk 3 above)
Comment (ssoloff) [inline build.install4j]: Derp.  Of course it does.  Otherwise, it wouldn't be much of an install tool, right? :smiley:

File: build.install4j (hunk 4 above)
Comment (ssoloff) [inline build.install4j]: It might be a new feature added in 6.1.5 and that's just the default value.  I can guess what it does, but we should try to research a bit more just in case it's something undesirable.

File: build.install4j (hunk 3 above)
Comment (RoiEXLab) [inline build.install4j]: Correct ^^

File: build.install4j (hunk 4 above)
Comment (RoiEXLab) [inline build.install4j]: I just tested it,
I loaded the current project file, added a single space to a string, removed it and saved the project.
This line was added because of that...

Comment (ssoloff) [inline build.install4j]: Found it.  It has to do with auto-starting installed services (i.e. registering them with `systemd`).  It appears in 6.1.5:

install4j-6 1 5-launcher-unix-options

But is not present in 6.1.2:

install4j-6 1 2-launcher-unix-options

So, it sounds like it's something we don't need to worry about.

File: build.install4j (hunk 5 above)
Comment (ssoloff) [inline build.install4j]: I ran the Windows x64 installer as a non-admin user.  I actually expected to get an elevation prompt but didn't see one.  Then I looked at `HKCR` to see if the `triplea` key was created, and I couldn't find it.  Here's the first registry-related error from the install log:

```
[INFO] com.install4j.runtime.beans.actions.registry.SetRegistryValueAction [ID 172]: Execute action
       Property value: URL:TripleA Protocol
       Property createKey: true
       Property valueName: 
       Property keyName: triplea
       Property registryRoot: HKEY_CLASSES_ROOT
       Property rollbackSupported: true
[ERROR] com.install4j.runtime.beans.actions.registry.SetRegistryValueAction [ID 172]: could not create key HKEY_CLASSES_ROOT triplea
        Execute action not successful after 1 ms
```

I will try rebuilding the installer with a root key of `HKCU\Software\Classes` instead of `HKCR` and see what happens.

File: build.install4j [hunk 6]
```diff
@@ -535,6 +706,72 @@
...
+                          <class>com.install4j.api.windows.RegistryRoot</class>
+                          <string>HKEY_CLASSES_ROOT</string>
+                        </object>
+                      </void>
+                    </object>
+                  </java>
+                </serializedBean>
+                <condition />
+              </action>
+              <action name="" id="221" customizedId="" beanClass="com.install4j.runtime.beans.actions.control.RunScriptAction" enabled="true" commentSet="false" comment="" actionElevationType="inherit" rollbackBarrier="false" multiExec="false" failureStrategy="1" errorMessage="Failed to remove protocol Handler">
```
Comment (ssoloff) [inline build.install4j]: Please change "Handler" to "handler" in the error message.

File: build.install4j [hunk 7]
```diff
@@ -334,6 +334,177 @@
...
+                          </void>
+                          <void property="value">
+                            <string>"${installer:sys.installationDir}\TripleA.exe" "%1"</string>
+                          </void>
+                        </object>
+                      </java>
+                    </serializedBean>
+                    <condition />
+                  </action>
+                  <action name="" id="220" customizedId="" beanClass="com.install4j.runtime.beans.actions.control.RunScriptAction" enabled="true" commentSet="false" comment="" actionElevationType="inherit" rollbackBarrier="false" multiExec="false" failureStrategy="1" errorMessage="Failed to create protocol Handler">
```
Comment (ssoloff) [inline build.install4j]: Please change "Handler" to "handler" in the error message.

File: build.install4j (hunk 5 above)
Comment (ssoloff) [inline build.install4j]: I built the Windows x64 installer with this [patch](https://github.com/triplea-game/triplea/files/1045703/pr-1783-protocol-handler-registry-keys.diff.txt) and installed it as a non-admin user.  I now see the `triplea` key and all its values under `HKCR`.  After entering the URL `triplea:test` into Firefox, I was prompted to run TripleA.  I observed no errors in the log.

**Comment by RoiEXLab** [comment]
@ssoloff It's a bit hard to follow all of your instructions... 
If you have the time: Could you create a PR on my fork to add the minor changes?

**Comment by RoiEXLab** [comment]
Future thought: Maybe we should create 2 installer versions/add an option to the installer prompting you to decide wether you want to install TripleA for the current user (no admin rights required) or all users (admin rights required).

**Comment by ssoloff** [comment]
> Could you create a PR on my fork to add the minor changes?

Sure, @RoiEXLab, I can do it in a few hours.  Can you please merge the latest changes from `master` first?

**Comment by RoiEXLab** [comment]
@ssoloff Done!

**Comment by simon33-2** [comment]
So does this mean you can open a tsvg file by double clicking?

**Comment by RoiEXLab** [comment]
@simon33-2 No, but I wanted to add this functionality anyway.
I will add this in a future PR

**Comment by simon33-2** [comment]
I'm confused about what this PR achieves then?

**Comment by RoiEXLab** [comment]
@simon33-2 Read https://github.com/triplea-game/triplea-game.github.io/issues/33#issuecomment-302188937 for more information

**Comment by ssoloff** [comment]
@RoiEXLab 

> 2 Questions:
>
>    Should we go for a flag like -o? in this case we'd call triplea like this: /path/to/triplea -o triplea:<somestring>

Please review this [comment](https://github.com/triplea-game/triplea/issues/1723#issuecomment-302281868).  If you look at `GameRunner#handleCommandLineArgs()`, you can see that TripleA currently assumes all command-line arguments are passed as `<key>=<value>` pairs, with the exception of the first argument which may omit the key and is assumed to be `triplea.game`.  I would assume we don't want to introduce an entirely new way to process the command line at this time (tagging @ron-murhammer and @DanVanAtta in case they have a different opinion).  So defining a new key and passing that on the command line is probably the way to go.

>    What are we going to pass in as params? just the maps name? The full URL? Both?

`DownloadMapsWindow` currently works with map name, so that would be easiest (it uses the map name to lookup an instance of `DownloadFileDescription`).  I think the map name can be considered a unique identifier for the map, so it should be ok.  Note that map names can contain spaces, so you'll have to make sure to surround your command line argument in quotes.

In the long run, I can see maybe supporting an URL if we want to download a map that's not in `triplea_maps.yaml` for whatever reason.  So, just to be safe, maybe make sure the key you choose for the command line argument specifically indicates it's a map name (e.g. `triplea.downloadMapName=<mapName>`).  Then, in the future, we can add `triplea.downloadMapUrl=<mapUrl>` if needed, and they can co-exist.

**Comment by RoiEXLab** [comment]
@ssoloff I changed the scripts to behave according to the scheme: "Installer overrides existing settings, uninstaller only removes changes that were introduced by the corresponding installer (same version)"
I haven't been able to test the code I wrote though...
I will do that later this day

**Comment by RoiEXLab** [comment]
@DanVanAtta @ssoloff Merge conflicts were resolved, installers were tested...
We should be good to go!
Current behaviour with parallel installs:
1. Install A: `triplea: -> A`
2. Install B: `triplea: -> B`
3. Uninstall A: `triplea: -> B`
4. Uninstall B: `triplea: -> none`
---
1. Install A: `triplea: -> A`
2. Install B: `triplea: -> B`
3. Uninstall B: `triplea: -> none`
4. Uninstall A: `triplea: -> none`

You might have noticed, that the protocol handler entry for A is not being restored in the second example. The reason I chose not to implement this kind of feature is because it would make stuff way too complicated because we would then need to keep track of parallel triplea installations and when they are removed... If anyone wants the protocol handler back, they can simply reinstall triplea.

I found the File Association feature btw... The next PR will regiter .tsvg files to be launched with triplea

File: build.install4j [hunk 8]
```diff
@@ -535,6 +699,70 @@
...
+                      <void property="registryRoot">
+                        <object class="java.lang.Enum" method="valueOf">
+                          <class>com.install4j.api.windows.RegistryRoot</class>
+                          <string>HKEY_CURRENT_USER</string>
+                        </object>
+                      </void>
+                    </object>
+                  </java>
+                </serializedBean>
+                <condition>Util.isWindows() &amp;&amp; ((String)WinRegistry.getValue(RegistryRoot.HKEY_CURRENT_USER, "Software\\Classes\\triplea\\shell\\open\\command", "")).equalsIgnoreCase("\"" + context.getVariable("sys.installationDir") + "\\TripleA.exe\" \"%1\"")</condition>
```
Comment (ssoloff) [inline build.install4j]: Just curious about two things...

1. Why was it necessary to add `Util.isWindows()` to the condition here but not in the `SetRegistryValueAction`s?  I would have assumed as with the `SetRegistryValueAction`s, install4j "knows" that `DeleteRegistryItemAction` can only run on Windows.  Is it because it is a compound condition and install4j will evaluate the second sub-expression regardless of the OS (and thus fail on non-Windows systems)?
1. Is the second sub-expression (`WinRegistry.getValue(...`) the magic that leaves a new installation's protocol handler intact when you uninstall an older version as you described [here](https://github.com/triplea-game/triplea/pull/1783#issuecomment-307525201)?

**Comment by simon33-2** [comment]
I can test on a mac but I don't have a licence key.

File: build.install4j (hunk 8 above)
Comment (RoiEXLab) [inline build.install4j]: 1. Yes, it is necessary, I got an UnsatisfiedLinkError on linux without this.
I believe this is a install4j bug. It seems that in thus case the condition is evaluated before install4j checks if the action is valid for the current os.
2. Yes, this makes the triplea installer only remove the entries if the protocol handler points at the current installation.

**Comment by RoiEXLab** [comment]
@simon33-2 just go to the install4j website and get yourself a 90 days trial key... It'll work

**Comment by simon33-2** [comment]
Then I wouldn't be able to do that later would I? Easier to have a build.

**Comment by RoiEXLab** [comment]
@simon33-2 I can send you the dmg file of course, hold on a couple minutes.

**Comment by RoiEXLab** [comment]
@ssoloff If you have the time feel free to test the installer on fedora. Its probably good to test the installer on non-debian based systems, although I'm not 100% sure if those are supported by install4j whrn it comes to create desktop file.

**Comment by RoiEXLab** [comment]
@simon33-2 @ssoloff Here you go: [releases.zip](https://expirebox.com/download/121cec98da1c95f2b4bc46e1b7c94b3f.html)
(Link will expire in 2 days, zip file contains a sh and a dmg file)

**Comment by RoiEXLab** [comment]
@simon33-2 OS X is not supported by this PR although the OS seems to support such a behaviour. Could you write a java script (using hardcoded values is enough, I can replace them) that registerst triplea as protocol handler?

File: build.install4j (hunk 8 above)
Comment (RoiEXLab) [inline build.install4j]: (For whatever reason I don't read your messages word for word, so here's the correct answer 😅)
Install4j knows that the registry actions can only be run on windows systems Therefore this is likely a bug.
It seems that install4j is first evaluating the condition, then checking if this action is appropriate for the current os and the execute it. The first 2 actions should be in reversed order.
So yes, your assumptions seems to be correct.

**Comment by ssoloff** [comment]
@RoiEXLab I built the installer directly from your PR and tested it on Fedora 25 (4.10.15-200.fc25.x86_64).  It apparently failed to install the protocol handler:

install-failure

The install log isn't very helpful:

```
[ERROR] com.install4j.runtime.beans.actions.control.RunScriptAction [ID 220]: Execute action not successful after 43 ms
```

Reviewing the installer action, one thing I noticed is that you return `false` if `desktopFile` does not exist:

```java
File desktopFile = new File(applicationsDir, "TripleA-" + context.getCompilerVariable("sys.version") + "-0.desktop");
File mimeappsFile = new File(applicationsDir, "mimeapps.list");
if(!desktopFile.exists()){
    return false;
}
```

This file does not exist in my `~/.local/share/applications` folder.  Should it exist prior to a TripleA installation?

I'm assuming when you return `false` from a script action, that's what triggers the install4j failure, correct?

**Comment by RoiEXLab** [comment]
@ssoloff Yes, that is correct. In this case it seems, that fedore is not supported by install4j as an operating system.
I never used fedora... Is something like a "start menu entry" like on windows created?

**Comment by ssoloff** [comment]
@RoiEXLab There is no "start menu entry" even though the `com.install4j.runtime.beans.actions.desktop.RegisterAddRemoveAction` says that it's successful.

I don't think it's a problem if we want to say Red Hat OS'es are not supported by this feature, but I also don't think the installer should display a failure message to the user in this case.  Instead of returning `false`, can we write a message to the log stating the OS isn't supported by the TripleA protocol handler feature and return `true`?

**Comment by RoiEXLab** [comment]
@ssoloff Changed it to return true and display a message if this file is not existing...

**Comment by ssoloff** [comment]
Excellent, @RoiEXLab!  Although, I should have explicitly stated that I was thinking of calling `Util#logInfo()` instead of displaying a dialog to the end user.  I'm not sure it's important enough to interrupt the installation and require the user to confirm by closing the dialog.  If you disagree, I'm ok with leaving it as-is.

One other thing I noticed in the latest round of testing is that even though the installer is not modifying `mimeapps.list` during installation on my system, it is modifying it during uninstallation.  My `mimeapps.list` file is originally empty.  After installing TripleA, it is still empty.  After uninstalling TripleA, it is one byte long containing just a single line feed (0xA).  It's not a big deal, but looking at the code again, we may be, in general, just writing an extra newline here:

```java
try(PrintWriter writer = new PrintWriter(mimeappsFile)){
    writer.println(builder.toString());
}
```

It looks like you are already appending a newline to every line you add to  the `StringBuilder`.  In that case, can you just use `writer.print()` instead of `writer.println()`?

**Comment by RoiEXLab** [comment]
> It looks like you are already appending a newline to every line you add to the `StringBuilder`. In that case, can you just use `writer.print()` instead of `writer.println()`?
@ssoloff Yes, of course! I fixed that

**Comment by ssoloff** [comment]
Tested again and everything looks fine.  Good job, @RoiEXLab!

Once the protocol handler stuff is working through the website and everything, I'm sure I'll want to investigate getting it to work on Red Hat. :smile:

**Comment by RoiEXLab** [comment]
@ssoloff Good luck with that...
Just a question:
Does fedora use mimeapps.list files? In this case we'd just need to create a .desktop file manually and we'd be good to go...

**Comment by ssoloff** [comment]
@RoiEXLab What I can tell you is I have a `~/.local/share/applications/mimeapps.list` file, but it is empty.  However, there are several `*.desktop` files in that folder, as well, for applications like Chrome (there seems to be one `.desktop` file per Chrome app I have installed) and Epiphany (the Fedora download manager).

**Comment by RoiEXLab** [comment]
@ssoloff Ok in this case install4j is the problem.
It should normally create a desktop file for triplea there which can be used by us to create the protocol handler...
Creating such a file manually should do the trick!

**Comment by RoiEXLab** [comment]
@ssoloff Or what would be even better:
You probably should contact the install4j support and point out that fedora is not supported although it could be by just creating files just like on Ubuntu or debian based systems in general

**Comment by ssoloff** [comment]
@RoiEXLab Good news...  Everything works fine on Fedora!

It was user error on my part.  I've been so used to installing TripleA as a dev, as opposed to a user, that I normally disable the step in the installer to "Create symlinks" (because I hate having to change `/usr/local/bin` to `/home/<user>/.local/bin` as I'm not running sudo).  It's this step that creates the `*.desktop` file.  So it's not that the action wasn't working, it just simply did nothing because I told it to. :smile:

After installing with symlinks enabled, I also verified that entering an address of `triplea:something` into Firefox launched TripleA.

Sorry to suggest another change, but I think it's clear now that the log message is unnecessary.  We're not failing to install the protocol handler because the OS is unsupported but rather because the user simply chose to not install symlinks.  Simply returning `true` is sufficient.

Maybe what I need to look into in the future is seeing if we can change the default symlink folder to the user's local binary folder rather than the system-wide folder.

**Comment by RoiEXLab** [comment]
@ssoloff Interesting, for me it was the other way round: After having installed triplea a couple times I had 2 open terminals to launch the (un)installers so all I had to do was to press arrow up enter next next next next do not run triplea reload triplea:test page in Firefox ^^
I will remove the message soon, we are waiting for @simon33-2's feedback anyways

**Comment by RoiEXLab** [comment]
I'd argue that since we tested this on various UNIX systems and on Windows, no regressions were introduced to mac OS...

**Comment by DanVanAtta** [comment]
Let's merge this up, looks like a good bit of testing has been done. @RoiEXLab , please do follow up to see if you can find some mac users to give this a try. We can roll back or update as appropriate if we find any problems.

> I'd argue that since we tested this on various UNIX systems and on Windows, no regressions were introduced to mac OS...

I would caution Mac can be different enough. Network config, ruby, apt-get vs brew, docker can all be slightly different. CLI can have some differences, example straight `sed -i` does not work on Mac OS.

**Comment by RoiEXLab** [comment]
@DanVanAtta Sorry, forgot to mention it...
Of course mac differs from linux in many ways! What I meant to say is that because we have `Util.isLinux()` and `Util.isWindows()` checks, no code should ever be executed on mac os unless install4j is broken!

**Comment by simon33-2** [comment]
@RoiExLab wrote: "@simon33-2 OS X is not supported by this PR although the OS seems to support such a behaviour. Could you write a java script (using hardcoded values is enough, I can replace them) that registers triplea as protocol handler?
I couldn't figure out how to do that just using thr little information I found."

Well my Java knowledge is pretty basic. If you give me a step by step to run on a Mac I can probably do it.

**Comment by RoiEXLab** [comment]
@simon33-2 I'm assuming that your mac already has java installed on it and java paths are configured correctly.
(You can easily check this by opening a Terminal and entering `java -version` and/or `javac -version`. If this commands prints the java(c) version everything is configured correctly.)
As we need just a simple script here, we don't need an IDE (although we could still use it to make things easier if we wanted to).
You create a File called `Main.java` with the following Code:
```
//Add your imports here
public class Main {
  public static void main(String[] args){
    //Here goes your code
  }
}
```
After you believe your code is finished, run `javac Main.java` to compile the code and afterwards `java Main` to run it.
If this is too complicated for you, you can of course just tell me how I need to edit which files in order to achieve this, I can then come up with a script myself, which you'd then need to test.
